src/defs.rs

```diff
@@ -53,6 +57,76 @@ pub struct WorkItem {
     pub name: String,
 }
 
+#[derive(Debug)]
+pub enum ArchiveType {
+    Zip(ZipArchive<File>),
+    Dir(PathBuf),
+}
+
+#[derive(Debug)]
+pub struct Archive {
+    pub item: ArchiveType,
+}
+
+impl Archive {
+
+    pub fn read_in_buffer(&mut self, name: &str, buf: &mut Vec<u8>) -> bool {
+        match self.item {
+            ArchiveType::Zip(ref mut zip) => {
+                match zip.by_name(name) {
+                    Ok(mut f) => {
+                        f.read_to_end(buf).expect("Failed to read gcda file");
+                        true
+                    },
+                    Err(_) => false,
+                }
+            },
+            ArchiveType::Dir(ref dir) => {
+                match File::open(dir.join(PathBuf::from(name))) {
```

Comment:
It might be possible to join with a string directly, instead of converting it to PathBuf first (I'm not sure).